Document a list of API differences

Srivaralakshmi opened the pull request:
[RHDEVDOCS-4370](https://issues.redhat.com/browse/RHDEVDOCS-4370): [SB] Document a list of features supported for the CoreOS and Spec APIs

- **Aligned team**: Dev Tools
- **OCP version for cherry-picking**: enterprise-`4.10`, enterprise-`4.11` and later
- **JIRA issues**: [RHDEVDOCS-4370](https://issues.redhat.com/browse/RHDEVDOCS-4370)
- **Preview pages**: [Download to see the preview in your browser](https://drive.google.com/file/d/19bBn8VewcI7i3RoqmRdR-gQA0qjDgG1m/view?usp=sharing)
- **SME Review**: Completed by @sadlerap, @dperaza4dustbit
- **QE review**: Completed by @pmacik 
- **Peer-review**: Completed by @gabriel-rh

Srivaralakshmi commented:
@sadlerap, @dperaza4dustbit, and @pmacik: PTAL at the PR and provide ffedback. Thanks!

Srivaralakshmi commented:
Note to reviewers: Need to review **only** the content about [API differences](https://github.com/openshift/openshift-docs/pull/50261/files#diff-6b12b04cc17180221f6fd718cfd8e0c07b913a5ca67c8e3544aca6ec2bc44ff8). The rest of the content is _already_ reviewed and is included here in the PR only for modularization purposes. Thanks!

gabriel-rh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,75 @@
+* *Service Binding* with the `binding.operators.coreos.com` API group.
+* *Service Binding (Spec API)* with the `servicebinding.io` API group.
+
+Both of these API groups have similar features, but they are not completely identical.  Here is the complete list of differences between these API groups:
```
```suggestion
Both of these API groups have similar features, but they are not completely identical. Here is the complete list of differences between these API groups:
```

Srivaralakshmi replied:
Done

gabriel-rh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,75 @@
+| Alternative binding sources (for example, binding data from annotations)
+| Yes
+| Allowed by {servicebinding-title}
+| The specification says that we need to support getting binding data from provisioned services and secrets.  However, by a strict reading of the spec, we are not disallowed from supporting other binding data sources. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these so ...
```
```suggestion
| The specification says that we need to support getting binding data from provisioned services and secrets. However, by a strict reading of the spec, we are not disallowed from supporting other binding data sources. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these sources on both the API groups.
```

Srivaralakshmi replied:
Done

gabriel-rh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,75 @@
+| Alternative binding sources (for example, binding data from annotations)
+| Yes
+| Allowed by {servicebinding-title}
+| The specification says that we need to support getting binding data from provisioned services and secrets.  However, by a strict reading of the spec, we are not disallowed from supporting other binding data sources. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these so ...
```
"we need to support"  and "we are not disallowed from supporting"

These should probably be re-worded to remove "we". Maybe say "The software"/"The API"/ etc or even change to passive (which should be ok when talking about an API).  So maybe something like:

"The specification requires support for getting binding data from provisioned services and secrets"

"However, a strict reading of the specification suggests that support for other binding data source is allowed"

Srivaralakshmi replied:
Hi @sadlerap: PTAL at the [suggestions](https://github.com/openshift/openshift-docs/pull/50261#discussion_r968612365) from Gabriel, the last 2 lines. Let me know whether it reads fine for you. Once you confirm, I will implement those changes.

Srivaralakshmi replied:
Done

gabriel-rh reviewed (commented):
Looks good! some minor whitespace issues and change from using "we".

Vale complains a bit about using the passive, but I think it is OK when talking about an API. 

It also suggests using "equal" rather than "equivalent", but I think the latter sounds better here so would keep it.

Srivaralakshmi commented:
Hi @gabriel-rh, Thanks for the suggestions. I have implemented them. PTAL at the revised PR and provide feedback.

Srivaralakshmi commented on modules/sbo-api-differences.adoc:
```diff
@@ -71,5 +71,5 @@ the workload.
 | Yes
 | Allowed by {servicebinding-title}
-| The specification says that we need to support getting binding data from provisioned services and secrets.  However, by a strict reading of the spec, we are not disallowed from supporting other binding data sources. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these so ...
+| The specification requires support for getting binding data from provisioned services and secrets. However, by a strict reading of the specification suggests that support for other binding data sources are allowed. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these sou ...
```
Done

gabriel-rh reviewed (approved):
Looks good to me!

Srivaralakshmi commented:
Thanks, @gabriel-rh. Can you please give me an `/lgtm`?

gabriel-rh commented:
/lgtm

maxwelldb commented on modules/sbo-about-service-binding-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+// * /applications/connecting_applications_to_services/understanding-service-binding-operator.adoc
+
+:_content-type: CONCEPT
+[id="sbo-about-service-binding-operator_{context}]
```
@Srivaralakshmi This ID isn't closed.

Srivaralakshmi replied:
Thanks @maxwelldb, Fixed it! Waiting for the build. Hope it resolves fine.

Srivaralakshmi commented:
Hi, @gabriel-rh: Regarding the PR-https://github.com/openshift/openshift-docs/pull/50261, Just made some updates to module file metadata, fixed an internal cross-reference, and added Anchoring IDs in module files. No other content changes. PTAL and provide your LGTM. Thanks!

gabriel-rh commented:
/lgtm

gabriel-rh reviewed (approved):
looks good

pmacik commented:
/lgtm

Srivaralakshmi commented:
@sadlerap and @dperaza4dustbit PTAL at this PR and provide your feedback. Thanks!

Srivaralakshmi commented:
/label dev-tools
/label peer-review-done

sadlerap reviewed (approved):
/lgtm

Srivaralakshmi commented:
/label merge-review-needed

rolfedh commented on modules/sbo-about-service-binding-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+[id="making-an-operator-managed-backing-service-bindable_{context}"]
+== Making an Operator-managed backing service bindable
+To make a service bindable, as an Operator provider you need to expose the binding data required by workloads to bind with the services provided by the Operator. You can provide the binding data either as annotations or as descriptors in the CRD of the Operator that manages the backing service.
```
```suggestion
To make a service bindable, as an Operator provider, you need to expose the binding data required by workloads to bind with the services provided by the Operator. You can provide the binding data either as annotations or as descriptors in the CRD of the Operator that manages the backing service.
```

rolfedh commented on modules/sbo-about-service-binding-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+* Bind your workloads to Operator-managed backing services.
+* Automate configuration of binding data.
+* Provide service operators a low-touch administrative experience to provision and manage access to services.
```
```suggestion
* Provide service operators with a low-touch administrative experience to provision and manage access to services.
```

rolfedh commented on modules/sbo-about-service-binding-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+* Bind your workloads to Operator-managed backing services.
+* Automate configuration of binding data.
+* Provide service operators a low-touch administrative experience to provision and manage access to services.
+* Enrich development lifecycle with a consistent and declarative service binding method that eliminates discrepancies in cluster environments.
```
```suggestion
* Enrich the development lifecycle with a consistent and declarative service binding method that eliminates discrepancies in cluster environments.
```

rolfedh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,80 @@
+| Yes
+a| * Default behavior for the service bindings of the `binding.operators.coreos.com` API group.
+* Opt-in functionality for the service bindings of the `servicebinding.io` API group:
+environment variables are created alongside files.
```
```suggestion
Environment variables are created alongside files.
```

rolfedh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,80 @@
+| Container path
+| Yes
+| Partial
+| Because a service binding of the `binding.operators.coreos.com` API group can specify mapping behavior within the `ServiceBinding` resource, the
+`servicebinding.io` API group cannot fully support an equivalent behavior without more information about
+the workload.
```
```suggestion
| Because a service binding of the `binding.operators.coreos.com` API group can specify mapping behavior within the `ServiceBinding` resource, the `servicebinding.io` API group cannot fully support an equivalent behavior without more information about the workload.
```

rolfedh replied:
- Please remove hard line breaks like these. I see multiple instances of them in this PR. The openshift-docs repo doesn't allow them.

rolfedh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,80 @@
+| Alternative binding sources (for example, binding data from annotations)
+| Yes
+| Allowed by {servicebinding-title}
+| The specification requires support for getting binding data from provisioned services and secrets. However, a strict reading of the specification suggests that support for other binding data sources are allowed. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these source ...
```
```suggestion
| The specification requires support for getting binding data from provisioned services and secrets. However, a strict reading of the specification suggests that support for other binding data sources is allowed. Using this fact, {servicebinding-title} can pull the binding data from various sources (for example, pulling binding data from annotations). {servicebinding-title} supports these sources on both the API groups.
```

rolfedh commented on modules/sbo-service-binding-terminology.adoc:
```diff
@@ -0,0 +1,16 @@
+[horizontal]
+Service binding:: The representation of the action of providing information about a service to a workload. Examples include establishing the exchange of credentials between a Java application and a database that it requires.
+Backing service:: Any service or software that the application consumes over the network as part of its normal operation. Examples include a database, a message broker, an application with REST endpoints, an event stream, an Application Performance Monitor (APM), or a Hardware Security Module (HSM).
+Workload (application):: Any process, running within a container. Examples include a Spring Boot application, a NodeJS Express application, or a Ruby on Rails application.
```
```suggestion
Workload (application):: Any process running within a container. Examples include a Spring Boot application, a NodeJS Express application, or a Ruby on Rails application.
```

rolfedh commented on modules/sbo-service-binding-terminology.adoc:
```diff
@@ -0,0 +1,16 @@
+Backing service:: Any service or software that the application consumes over the network as part of its normal operation. Examples include a database, a message broker, an application with REST endpoints, an event stream, an Application Performance Monitor (APM), or a Hardware Security Module (HSM).
+Workload (application):: Any process, running within a container. Examples include a Spring Boot application, a NodeJS Express application, or a Ruby on Rails application.
+Binding data:: Information about a service that you use to configure the behavior of other resources within the cluster. Examples include credentials, connection details, volume mounts, or secrets.
+Binding connection:: Any connection that establishes an interaction between the connected components such as a bindable backing service and an application requiring that backing service.
```
```suggestion
Binding connection:: Any connection that establishes an interaction between the connected components, such as a bindable backing service and an application requiring that backing service.
```

rolfedh commented on modules/sbo-about-service-binding-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+[id="binding-a-workload-together-with-a-backing-service_{context}"]
+== Binding a workload together with a backing service
+By using the {servicebinding-title}, as an application developer, you need to declare the intent of establishing a binding connection. You must create a `Service Binding` CR  that references the backing service. This action triggers the {servicebinding-title} to project the exposed binding data into the workload. The {servicebinding-title} receives the declared intent and binds the workload toget ...
```
I think custom resource names aren't allowed to contain space characters. Do you mean 
```
create a `ServiceBinding` CR
```
or 
```
create a service binding CR
```

rolfedh commented on modules/sbo-api-differences.adoc:
```diff
@@ -0,0 +1,76 @@
+
+The CRD of the {servicebinding-title} supports the following APIs:
+
+* *Service Binding* with the `binding.operators.coreos.com` API group.
+* *Service Binding (Spec API)* with the `servicebinding.io` API group.
```
I'm not sure about these bold references to service binding. I believe I commented on this issue in another PR I just reviewed for you. Please apply the same changes that you applied to that PR here in this PR and in any other similar instances you're working on. Thanks!

rolfedh commented:
@Srivaralakshmi this branch needs to be rebased.

rolfedh commented:
/cherry-pick enterprise-4.10

rolfedh commented:
/cherry-pick enterprise-4.12
